src/ptr.rs

```diff
@@ -0,0 +1,331 @@
+use anyhow::anyhow;
+#[cfg(not(target_arch = "wasm32"))]
+use proptest_derive::Arbitrary;
+use std::fmt::{Display, Formatter};
+use std::hash::Hash;
+use std::{fmt, marker::PhantomData};
+
+#[cfg(not(target_arch = "wasm32"))]
+use proptest::prelude::*;
+
+use libipld::Cid;
+use libipld::Multihash;
+
+use crate::light_data::Encodable;
+use crate::light_data::LightData;
+
+use crate::field::{FWrap, LurkField};
+use crate::tag::{ContTag, ExprTag, Tag};
+
+use serde::Deserialize;
+use serde::Serialize;
+use serde::{de, ser};
+
+use crate::hash::IntoHashComponents;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum RawPtr {
+    Null,
+    Opaque(usize),
+    Index(usize),
+}
+
+impl RawPtr {
+    pub fn new(p: usize) -> Self {
+        Self::Index(p)
+    }
+
+    pub const fn is_opaque(&self) -> bool {
+        matches!(self, Self::Opaque(_))
+    }
+
+    pub fn is_null(&self) -> bool {
+        matches!(self, Self::Null)
+    }
+
+    pub fn opaque_idx(&self) -> Option<usize> {
+        match self {
+            Self::Opaque(x) => Some(*x),
+            _ => None,
+        }
+    }
+
+    pub fn idx(&self) -> Option<usize> {
+        match self {
+            Self::Index(x) => Some(*x),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct Ptr<F: LurkField> {
+    pub tag: ExprTag,
+    pub raw: RawPtr,
+    pub _f: PhantomData<F>,
+}
+
+#[allow(clippy::derive_hash_xor_eq)]
+impl<F: LurkField> Hash for Ptr<F> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.tag.hash(state);
+        self.raw.hash(state);
+    }
+}
+
+impl<F: LurkField> Ptr<F> {
+    // TODO: Make these methods and the similar ones defined on expression consistent, probably including a shared trait.
```

Comment:
Wondering if we should mirror the `SPtr<E: Tag, F: LurkField>` + type-alias technique for `Ptr` and `ContPtr`

Reply:
Yes. Please open an issue.